Retry handlers fix

cstamas opened the pull request:
Seems that Hc4Provider misses to set retry handler (default is 3).

But since AbstractProxy class also handler retries, we actually get 3x3 retries?

(seen in RSO logs)

cstamas committed:
Retries handling was missed in Hc4Provider

cstamas committed:
Moving stuff to proper place.

jdillon commented:
Aside from all the ancillary changes in this pull... +1

peterlynch commented:
-1 please add a unit test

jdillon commented:
@peterlynch the main change her is configuring StandardHttpRequestRetryHandler (an hc4 class) on the HttpClient instance... what exactly would a unit test be testing?

peterlynch commented:
@jdillon that values from getRetrievalRetryCount() are set as the number of retries

jdillon commented:
ic, so the passing of global config on to httpclient, okay.

cstamas committed:
UT for Retry handler fix

cstamas commented:
Rebased against latest, added UT that tests Hc4Provider (is retryCount properly applied?), and test for HttpClientManager (is retryCount properly nullified, as these instances are used in RRS)

peterlynch commented on nexus/nexus-proxy/src/test/java/org/sonatype/nexus/proxy/storage/remote/httpclient/HttpClientManagerTest.java:
```diff
@@ -99,4 +128,21 @@ public void doNotFollowRedirectsToDirIndex()
+                globalRemoteStorageContext );
+        // check is all set as needed: retries should be not attempted, as it is manually handled in proxy repo
+        Assert.assertTrue( ( (DefaultHttpClient) client ).getHttpRequestRetryHandler() instanceof StandardHttpRequestRetryHandler );
+        Assert.assertTrue( globalRemoteStorageContext.getRemoteConnectionSettings().getRetrievalRetryCount() != 0 );
```
would prefer use hamcrest or Fest Asserts

In future avoid using assertTrue/assertEquals when testing measurable conditions so that when it fails, we have a more meaningful message than something wasn't true. Using assertThat in this case could for example tell us what it was equal to or instance of.

peterlynch commented:
+1 thanks for tests

cstamas commented:
CI
https://builds.sonatype.org/view/nexus-features/job/nexus-oss-its-feature/488/

cstamas commented:
Not merging, CI shows some problems with content validation ITs.... will check out what happened here

cstamas committed:
Undoing the actual fix, and leaving the problem in
place (actual retries will be repoRetries squared)

cstamas commented:
Rebased and undid the fix. The bug is still there, even if the IT will pass. It will pass as HC4 will retry more then once (on it's own, using it's own retry handler) as 1st times it will fail. IT log will be full of NoHttpResponseExceptions.

cstamas commented:
Recorded issue NEXUS-5368
